### DISCUSSION TITLE
eigenstrain about two variables

### DISCUSSION THREAD

**Discussion by wangzhaohao** [idaholab/moose#23510] [Q&A General]
I want to self edite a eigenstrain. So i see the `ComputeThermalExpansionEigenstrainBase` and `ComputeThermalExpansionEigenstrain`. i have a question about the _deigenstrain_dT. The strian formaltion is that
$\epsilon_D = \triangle \rho (e^{\frac{T}{B}})$
in there the variables are T and B
i don't konw the _deigenstrain_dT is only derivative about T or about T and B. If later, i should write the `addIa(thermal_strain.derivatives()) and `addIa(thermal_strain.derivatives())(which is about B)`?
```C++
// in ComputeThermalExpansionEigenstrainBase.C
_deigenstrain_dT(is_ad ? nullptr 
                       : &this->tempalte declarePropertyDerivative<RankTwoTensor>(
                       "_eigenstrain_name, this->getVar("temperature", 0)->name()))

(*_deigenstrain_dT)[_qp].addIa(thermal_strain.derivatives());
// in myself define
_deiginstrain_dT...// same like above
_deigenstrain_dB....// same like above
(*_deigenstrain_dT)[_qp].addIa.//xxxxx(i can't)
```
I think my question core is that i don't farmaily with deigenstrain (what's this doing in MOOSE? to construct the Jacobian?)
thanks for your help!
zhaohao Wang

**Comment by GiudGiud** [comment]
Hello

You can use the "coupled()" routine on temperature to get the index of temperature. Then you should be able to index into derivatives() to retrieve the derivative with regards to temperature. You ll need to get the degree of freedom (dof) index, this can be found from the element or the node. 
Either that or if you have a pointer to the variable (using getVar in the constructor for example) then there are "number()" routines.

An example of doing that:
```
    const Elem * const sided_elem = ssf.face_side;
    const auto dof_number = sided_elem->dof_number(_sys.number(), _var.number(), 0);
    const auto rho_face = _rho(ssf);
    const Real d_rho_face_d_dof = rho_face.derivatives()[dof_number];
```
Guillaume

**Reply by wangzhaohao** [reply]
but there are two variables. your mean is that i only need to care about T? don't need B?

**Reply by GiudGiud** [reply]
No I mean that both derivatives are there when you call derivatives() so you need to pick the right one for setting deigenstrain_dT.

Same for deigenstrain_dB

**Reply by wangzhaohao** [reply]
is there are some example for this? 
I think if there have  similar **Material** will help me a lot.

**Reply by GiudGiud** [reply]
not that I know but feel free to grep the repository for "derivatives()"

**Reply by hugary1995** [reply]
`derivatives()` is a [MetaPhysicL](https://github.com/roystgnr/MetaPhysicL) routine for retrieving derivatives of a dual number in forward mode automatic differentiation.

For the common AD use case where you are interested in the derivative of a function `g(f)` with a known variable `f` and its known derivatives `df/dx`, forward mode AD gives you the dual number `(g, dg/dx)` by evaluating `g(f)`.

The code snippet you showed above is a more advanced usage of MetaPhysicL's forward AD. For the more advanced use case, we are using forward mode AD to also compute the second derivative.

The comment in the source file does a good job at explaining what's happening:
```
  /**
   * computeThermalStrain must be overridden in derived classes. The return type
   * ValueAndDerivative<is_ad> contains the value for the thermal strain and its
   * temperature derivative. Derived classes should use `_temperature[_qp]` to obtain
   * the current temperature. In the is_ad == false case that member variable is
   * agumented and will be of the type ChainedReal. I.e. even with is_ad == false
   * a variant of forward mode automatic differentiation will be used internally to
   * compute the thermal strain and no manual implementation of the temperature derivative
   * is needed.
   */
```

However, downstream developers are responsible for overriding the compute eigenstrain method to also compute the derivative of the eigenstrain w.r.t. other variables, if any. Derivatives of the eigenstrain w.r.t. other variables will eventually appear in the Jacobian as you said.

**Reply by wangzhaohao** [reply]
So i use **ComputeVariableEigenstrain** is OK?

**Reply by wangzhaohao** [reply]
or can i use this
```C++
_deigenstrain_dT[_qp].addIa(instantaneous_cte_T)
_deigenstrain_dB[_qp].addIa(instantaneous_cte_B)
```
